### ISSUE TITLE
Rule Change: [no-loop-func] Ignore calls to Array methods (and possibly other functions that don’t escape the passed closure)

### ISSUE THREAD

**Issue by j-f1** [eslint/eslint#16953]
### What rule do you want to change?

no-loop-func

### What change to do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```js
for (var x = 0; x < 10; x++) {
  console.log(numbers.filter(n => n % x === 0))
}
```

### What does the rule currently do for this code?

It says: “Function declared in a loop contains unsafe references to variable `x`.”

### What will the rule do after it's changed?

Not warn here because the function is discarded before the loop runs again. 

### Participation

- [X] I am willing to submit a pull request to implement this change.

### Additional comments

I’m not sure how this check could be implemented — maybe this is would be a better candidate for TS-ESLint? It would need to somehow know that the function in question will not keep the callback it’s been passed around beyond its execution scope.

**Comment by mdjermanovic** [comment]
This was discussed in https://github.com/eslint/eslint/issues/5044 and the conclusion was that we cannot know that any method named `filter` is safe.

> maybe this is would be a better candidate for TS-ESLint

It might be a good idea to submit this proposal to `typescript-eslint` as there's already extension rule `@typescript-eslint/no-loop-func` and it might be able to use typo info to determine that this is, for example, `Array.prototype.filter`, which is safe in this context since it runs the callback right away and doesn't store a reference.
